### PR TITLE
Add option to disable Auto Merge + Add option to not check out tags + Add option to apply fetch settings to branches

### PR DESCRIPTION
**Pull request by DamienBitrise**
1. Not merging to master before building can significantly speed up the git clone. ie. git merge origin/master

2. Added the ability to optionally disable checking out tags 

3. Added the ability to apply clone depth and disable checkout tags to branches

New Settings:

image

All settings default to current functionality

This means when the step is updated nothing will change unless users change these 3 settings

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -125,7 +126,7 @@ func mainE() error {
 			return fmt.Errorf("checkout (%s): %v", checkoutArg, err)
 		}
 		// Update branch: 'git fetch' followed by a 'git merge' is the same as 'git pull'.
-		if checkoutArg == cfg.Branch {
+		if checkoutArg == cfg.Branch && cfg.AutoMerge{
```
Comment (DamienBitrise) [inline main.go:129]: Not sure if we should ever do merge origin/master after a fetch and checkout?

Seems to work fine

File: git.go [hunk 2]
```diff
@@ -226,12 +226,18 @@ func autoMerge(gitCmd git.Git, mergeBranch, branchDest, buildURL, apiToken strin
...
 
-func manualMerge(gitCmd git.Git, repoURL, prRepoURL, branch, commit, branchDest string) error {
+func manualMerge(gitCmd git.Git, repoURL, prRepoURL, branch, commit, branchDest string, autoMerge bool) error {
 	if err := runWithRetry(func() *command.Model { return gitCmd.Fetch("origin", branchDest) }); err != nil {
 		return fmt.Errorf("fetch failed, error: %v", err)
 	}
-	if err := pull(gitCmd, branchDest); err != nil {
-		return fmt.Errorf("pull failed (%s), error: %v", branchDest, err)
+	if autoMerge {
+		if err := pull(gitCmd, branchDest); err != nil {
```
Comment (DamienBitrise) [inline git.go:234]: pull does a checkout + merge

File: git.go [hunk 3]
```diff
@@ -226,12 +226,18 @@ func autoMerge(gitCmd git.Git, mergeBranch, branchDest, buildURL, apiToken strin
...
+			return fmt.Errorf("pull failed (%s), error: %v", branchDest, err)
+		}
+	} else {
+		if err := run(gitCmd.Checkout(branchDest)); err != nil {
```
Comment (DamienBitrise) [inline git.go:238]: We only need to do a checkout here

File: git.go [hunk 4]
```diff
@@ -226,16 +226,18 @@ func autoMerge(gitCmd git.Git, mergeBranch, branchDest, buildURL, apiToken strin
...
 	var opts []string
-	if !fetchTags {
-		opts = append(opts, "--no-tags")
-	}
-	if depth != 0 {
-		opts = append(opts, "--depth="+strconv.Itoa(depth))
-	}
-	if isTag {
-		opts = append(opts, "--tags")
+	if optionsOnBranches {
```
Comment (damo1884) [inline git.go:231]: This keeps the default behavior as is unless the setting is changed

Comment (DamienBitrise) [inline git.go:231]: This keeps the default behavior unless the setting is enabled

File: git.go [hunk 5]
```diff
@@ -269,7 +271,11 @@ func manualMerge(gitCmd git.Git, repoURL, prRepoURL, branch, commit, branchDest
 			return fmt.Errorf("merge failed (fork/%s), error: %v", branch, err)
 		}
 	} else {
-		opts = opts[:len(opts) - 2]
+		if optionsOnBranches {
```
Comment (DamienBitrise) [inline git.go:274]: This keeps the default behavior unless the setting is enabled

**Comment by BirmacherAkos** [comment]
Hey @DamienBitrise  👋 

Thanks for the features. I'd have a request and question related to the PR.

1. Could you please separate these 3 feature changes into 3 pull requests? I know it's a pain in the neck, but it would make it easier for us to review the changes. Other than that, if there will be a change request for one of the features, then we still could release the others.

2. The 3rd input is not completely clear to me. What will it do if I enable it? What happens if I use the 2. and the 3. inputs as well?

**Comment by DamienBitrise** [comment]
All 3 PRs will have merge conflicts so going to do one at a time

1. Auto Merge origin/master PR: https://github.com/bitrise-steplib/steps-git-clone/pull/114

**Comment by DamienBitrise** [comment]
Abandoning

**Pull request closed without merging**